### PR TITLE
Add Writer Version 4 to PROTOCOL.md

### PR DESCRIPTION
**Pull request by zsxwing**
- Add Writer Version 4 requirements.
- Add Generated Columns.
- Make the `Writer Version Requirements` table use GitHub markdown format.
- Update the table of contents.

### CODE REVIEW

File: PROTOCOL.md [hunk 1]
```diff
@@ -416,25 +426,25 @@ When the table property `delta.appendOnly` is set to `true`:
...
-+------------------+----------------------------------------------+
-| Writer Version 3 |  - Enforce:                                  |
-|                  |    - `delta.checkpoint.writeStatsAsJson`     |
-|                  |    - `delta.checkpoint.writeStatsAsStruct`   |
-|                  |    - `CHECK` constraints                     |
-+------------------+----------------------------------------------+
+<br> | Reader Version 1
+-|-
+Writer Version 2 | - Support [`delta.appendOnly`](#append-only-tables)<br>- Support [Column Invariants](#column-invariants)
+Writer Version 3 | Enforce:<br>- `delta.checkpoint.writeStatsAsJson`<br>- `delta.checkpoint.writeStatsAsStruct`<br>- `CHECK` constraints
```
Comment (zsxwing) [inline PROTOCOL.md:446]: `<br>` is to start a new line in a table cell.

File: PROTOCOL.md [hunk 2]
```diff
@@ -416,25 +426,25 @@ When the table property `delta.appendOnly` is set to `true`:
...
- - This column SHOULD be parsed as a boolean SQL expression.
+ - The `metadata` for a given column in the table schema MAY contain the key `delta.invariants`.
+ - This value of `delta.invariants` SHOULD be parsed as a boolean SQL expression.
  - Writers MUST abort any transaction that adds a row to the table, where a present invariant evaluates to `false` or `null`.
 
+## Generated Columns
+
+ - The `metadata` for a given column in the table schema MAY contain the key `delta.generationExpression`.
+ - This value of `delta.generationExpression` SHOULD be parsed as a SQL expression.
+ - Writers MUST enforce any data writing to the table satisfy the condition `(<value> <=> <generation expression>) IS TRUE`.
```
Comment (jaceklaskowski) [inline PROTOCOL.md:437]: nit: enforce **that** any...

File: PROTOCOL.md (hunk 1 above)
Comment (jaceklaskowski) [reply PROTOCOL.md:446]: I've been searching for this html / markdown trick for new lines in tables for so long. Thank you @zsxwing !

**Review by jaceklaskowski — APPROVED** [review]
LGTM

File: PROTOCOL.md [hunk 3]
```diff
@@ -416,25 +426,25 @@ When the table property `delta.appendOnly` is set to `true`:
...
- - Writers MUST abort any transaction that adds a row to the table, where a present invariant evaluates to `false` or `null`.
+ - The `metadata` for a column in the table schema MAY contain the key `delta.invariants`.
+ - The value of `delta.invariants` SHOULD be parsed as a boolean SQL expression.
+ - Writers MUST abort any transaction that adds a row to the table, where an invariant evaluates to `false` or `null`.
+
+## Generated Columns
+
+ - The `metadata` for a column in the table schema MAY contain the key `delta.generationExpression`.
+ - The value of `delta.generationExpression` SHOULD be parsed as a SQL expression.
+ - Writers MUST enforce that any data writing to the table satisfy the condition `(<value> <=> <generation expression>) IS TRUE`.
```
Comment (tdas) [inline PROTOCOL.md:437]: clarify that `<=>` is null safe equals, that is specify the null behavior

**Pull request closed without merging**